README
Games::Checkers is a set of Perl classes implementing the Checkers game
play. 17 different national rule variants (and any custom mix of rules)
are supported on any arbitrary board sizes. A basic AI heuristics is
implemented using the Minimax algorithm. Replay of previously recorded
games is supported too.

This package is intended to provide complete infrastructure for interactive
and automatic playing and manipulating of Checkers games.

See doc/rules to read about the Checkers game rules.
See doc/variants to find the rule differences in all supported variants.
Alternatively see Games::Checkers man page after installing this package.
You can run perl scripts in bin/ directory without installing anything.

Currently supported checkers/draughts variants (AI and game replay):

	* russian
	* russian_give_away
	* russian_10x8 (spantsiretti)
	* international (polish)
	* english
	* english_give_away
	* italian
	* spanish
	* argentinian
	* portuguese
	* czech
	* german
	* thai
	* pool (american_pool)
	* brazilian
	* frisian
	* canadian
	* sri_lankian

Currently supported board sizes:

	* 4x4, 6x6, 8x10 (just for fun)
	* 8x8
	* 10x8
	* 10x10
	* 12x12
	* 14x14, 16x16 (not too practical)

Note that every variant configures its own board size, but it is made
possible to play using any variant rules on different board sizes too,
for example Russian Checkers on 12x12 board. Or even construct own
rule variants, like: Give-Away International Checkers on the 8x10 board
without the requirement to capture maximal amount of pieces.

Currently supported game file formats:

	* .pdn files (trying to detect a lot of broken notations too)
	* .pdn.gz|xz|bz2 files (automatically uncompressed on the fly)

Currently installed scripts:

	* pcheckers-auto-play
	* pcheckers-replay-games

In the future one script pcheckers may be installed that will include:

	* automatic AI game play (current script pcheckers-auto-play)
	* recorded game replay (current script pcheckers-replay-games)
	* interactive game play of 1 or 2 human players

If SDL perl modules are installed, then the SDL support is automatically
detected and activated that replaces the default terminal IO. 

See also files INSTALL and TODO, and directory data/.

Author:
	Mikhael Goikhman <migo@freeshell.org>
Web site:
	http://migo.sixbit.org/software/pcheckers/


CHANGES
2018-03-22 - version 0.3.0
--------------------------

* Support for game variants (AI and game replay):
	* russian
        * russian_give_away
        * russian_10x8
        * international
        * english
        * english_give_away
        * italian
        * spanish
        * argentinian
        * portuguese
        * czech
        * german
        * thai
        * pool
        * brazilian
        * frisian
        * canadian
        * sri_lankian
* Support for variant aliases:
	* default   -> russian
	* british   -> english
	* internt   -> international
	* int       -> international
	* polish    -> international
	* 8x8       -> brazilian
	* 10x10     -> international
	* 12x12     -> canadian
	* shashki   -> russian
	* poddavki  -> russian_give_away
	* american  -> english
	* give_away -> english_give_away
	* spantsiretti  -> russian_10x8
	* american_pool -> pool
	* spanish_pool  -> spanish
	* 20 -> international
	* 21 -> english
	* 22 -> italian
	* 23 -> pool
	* 24 -> spanish
	* 25 -> russian
	* 26 -> brazilian
	* 27 -> canadian
	* 28 -> portuguese
	* 29 -> czech
	* 31 -> thai
	* 40 -> frisian
	* 41 -> russian_10x8
* Support for individual rule items:
	* BOARD_SIZE
	* BOARD_NOTATION
	* CAPTURE_SEPARATOR
	* STARTING_ROWS
	* BOTTOM_LEFT_CELL
	* PDN_GAME_TYPE
	* WHITE_STARTS
	* KINGS_LONG_RANGED
	* PAWNS_CAPTURING_BACKWARDS
	* CAPTURING_LARGEST
	* CAPTURING_POSTPONED
	* CROWNING_DURING_CAPTURE
	* PAWNS_CANT_CAPTURE_KINGS
	* CAPTURING_LEAVES_NO_GAP
	* GIVE_AWAY
	* CAPTURING_IN_8_DIRECTIONS
* Increased default AI depth to 5 from 3.
* Initial GUI menu in auto-play, to select game & board parameters.
* Support for editing initial board in GUI auto-play.
* New buttons in GUI in addition to "pause": "restart", "fullscreen".
* Better support for unusual cases in PDN format.
* Now all supplied pdn files are replayable.
* Refactored and improved code.
* Fixed build caused by bug in ExtUtils-MakeMaker & File-Find.


2012-05-13 - version 0.2.0
--------------------------

* Reworked code with lots of improvements.
* Support for GUI using SDL. Fallback to terminal output.
* Support for game file formats: .pdn, .pdn.gz, .pdn.bz2, .pdn.xz.
* Pausing game in GUI using Spacebar or icon.
* Moved to github repository mig0/Games-Checkers.


2004-03-01 - version 0.1.0
--------------------------

* Initial Checkers version with terminal output.
* Automatic AI game play, via script pcheckers-auto-play.
* Recorded game replay, via script pcheckers-replay-games.
* Support for boards:
	8x8
* Support for game variants (AI for "russian" only, game replay):
	* russian
	* british
	* italian
* Support for game file formats: .pdn, .pdn.gz.



lib/Games/Checkers/Rules.pm
# Games::Checkers, Copyright (C) 1996-2012 Mikhael Goikhman, migo@cpan.org
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

use strict;
use warnings;

package main;

our %RULES;

package Games::Checkers::Rules;

our %variant_rules = (
	international => {
		BOARD_SIZE                => '10x10',
		BOARD_NOTATION            => 'TL',
		CAPTURE_SEPARATOR         => 'x',
		STARTING_ROWS             => 4,
		BOTTOM_LEFT_CELL          => 1,
		PDN_GAME_TYPE             => 20,
		WHITE_STARTS              => 1,
		KINGS_LONG_RANGED         => 1,
		PAWNS_CAPTURING_BACKWARDS => 1,
		CAPTURING_LARGEST         => 1,
		CAPTURING_POSTPONED       => 1,
		CROWNING_DURING_CAPTURE   => 0,
		PAWNS_CANT_CAPTURE_KINGS  => 0,
		CAPTURING_LEAVES_NO_GAP   => 0,
		GIVE_AWAY                 => 0,
		CAPTURING_IN_8_DIRECTIONS => 0,
	},
	russian => {
		base => 'international',
		BOARD_SIZE                => '8x8',
		BOARD_NOTATION            => 'A1',
		CAPTURE_SEPARATOR         => ':',
		STARTING_ROWS             => 3,
		PDN_GAME_TYPE             => 25,
		CAPTURING_LARGEST         => 0,
		CROWNING_DURING_CAPTURE   => 1,
	},
	russian_give_away => {
		base => 'russian',
		PDN_GAME_TYPE             => 00,
		GIVE_AWAY                 => 1,
	},
	russian_10x8 => {
		base => 'russian',
		BOARD_SIZE                => '10x8',
		PDN_GAME_TYPE             => 41,
	},
	english => {
		base => 'international',
		BOARD_SIZE                => '8x8',
		BOARD_NOTATION            => 'TL',
		STARTING_ROWS             => 3,
		PDN_GAME_TYPE             => 21,
		WHITE_STARTS              => 0,
		KINGS_LONG_RANGED         => 0,
		PAWNS_CAPTURING_BACKWARDS => 0,
		CAPTURING_LARGEST         => 0,
		CAPTURING_POSTPONED       => 0,
		CROWNING_DURING_CAPTURE   => 'S',
	},
	english_give_away => {
		base => 'english',
		PDN_GAME_TYPE             => 00,
		GIVE_AWAY                 => 1,
	},
	italian => {
		base => 'english',
		BOARD_NOTATION            => 'TR',
		BOTTOM_LEFT_CELL          => 0,
		PDN_GAME_TYPE             => 22,
		WHITE_STARTS              => 1,
		CAPTURING_LARGEST         => 2,
		PAWNS_CANT_CAPTURE_KINGS  => 1,
	},
	spanish => {
		base => 'english',
		BOARD_NOTATION            => 'BR',
		BOTTOM_LEFT_CELL          => 0,
		PDN_GAME_TYPE             => 24,
		WHITE_STARTS              => 1,
		KINGS_LONG_RANGED         => 1,
		CAPTURING_LARGEST         => 3,
		CAPTURING_POSTPONED       => 1,
	},
	argentinian => {
		base => 'spanish',
		CAPTURING_LEAVES_NO_GAP   => 1,
		PDN_GAME_TYPE             => 00,
	},
	portuguese => {
		base => 'spanish',
		BOTTOM_LEFT_CELL          => 1,
		PDN_GAME_TYPE             => 28,
	},
	czech => {
		base => 'portuguese',
		BOARD_NOTATION            => 'A1',
		CAPTURING_LARGEST         => -1,
		PDN_GAME_TYPE             => 29,
	},
	german => {
		base => 'czech',
		CAPTURING_LARGEST         => 0,
		PDN_GAME_TYPE             => 00,
	},
	thai => {
		base => 'english',
		BOARD_NOTATION            => 'BR',
		CAPTURE_SEPARATOR         => '-',
		STARTING_ROWS             => 2,
		PDN_GAME_TYPE             => 31,
		KINGS_LONG_RANGED         => 1,
		CAPTURING_LEAVES_NO_GAP   => 1,
	},
	pool => {
		base => 'english',
		BOARD_NOTATION            => 'BR',
		PDN_GAME_TYPE             => 23,
		KINGS_LONG_RANGED         => 1,
		CROWNING_DURING_CAPTURE   => 0,
		PAWNS_CAPTURING_BACKWARDS => 1,
	},
	brazilian => {
		base => 'international',
		BOARD_SIZE                => '8x8',
		BOARD_NOTATION            => 'A1',
		STARTING_ROWS             => 3,
		PDN_GAME_TYPE             => 26,
	},
	frisian => {
		base => 'international',
		PDN_GAME_TYPE             => 40,
		CAPTURING_IN_8_DIRECTIONS => 1,
	},
	canadian => {
		base => 'international',
		BOARD_SIZE                => '12x12',
		STARTING_ROWS             => 5,
		PDN_GAME_TYPE             => 27,
	},
	sri_lankian => {
		base => 'canadian',
		BOTTOM_LEFT_CELL          => 0,
		PDN_GAME_TYPE             => 00,
	},
	default => 'russian',
	british => 'english',
	internt => 'international',
	int     => 'international',
	polish  => 'international',
	'8x8'   => 'brazilian',
	'10x10' => 'international',
	'12x12' => 'canadian',
	shashki   => 'russian',
	poddavki  => 'russian_give_away',
	american  => 'english',
	give_away => 'english_give_away',
	spantsiretti  => 'russian_10x8',
	american_pool => 'pool',
	spanish_pool  => 'spanish',
	20 => 'international',
	21 => 'english',
	22 => 'italian',
	23 => 'pool',
	24 => 'spanish',
	25 => 'russian',
	26 => 'brazilian',
	27 => 'canadian',
	28 => 'portuguese',
	29 => 'czech',
	31 => 'thai',
	40 => 'frisian',
	41 => 'russian_10x8',
);

sub set_variant ($%) {
	my $name = shift || 'default';
	my %params = @_;

	# support GameType specification from PDN
	my ($base, $starting_color, $size_x, $size_y, $notation, $invert) =
		split(',', $name);
	$params{WHITE_STARTS} = $starting_color eq 'W' ? 1 : 0
		if defined $starting_color;
	$params{BOARD_SIZE} = "${size_x}x$size_y"
		if $size_x && $size_y && "$size_x$size_y" =~ /^\d+$/;
	$params{BOARD_NOTATION} = $1 eq 'N' ? [ qw(BL BR TL TR) ]->[$2] : "$1$2"
		if defined $notation && $notation =~ /^([ANS])([0-3])$/;
	$params{BOTTOM_LEFT_CELL} = $invert ? 0 : 1
		if defined $invert;

	%::RULES = (variant => $base, base => $base);
	for (keys %{$variant_rules{international}}) {
		$::RULES{$_} = defined $params{$_} ? $params{$_} : $ENV{$_};
	}

	while (my $base = delete $::RULES{base}) {
		$base = lc($base); $base =~ s/[\s-]+/_/g;
		my $rules = $variant_rules{$base} || die "Checkers variant '$base' is unknown\n";
		$rules = { base => $rules } unless ref($rules);
		for (keys %$rules) {
			$::RULES{$_} = $rules->{$_} unless defined $::RULES{$_};
		}
	}

	return 1;
}

sub get_main_variants () {
	return map {
		my $entry = $variant_rules{$_};
		ref($entry) eq 'HASH' && defined $entry->{PDN_GAME_TYPE}
			? ($_) : ()
	} keys %variant_rules;
}

set_variant('default');

1;
